src/src/PluginStructs/P126_data_struct.cpp

```diff
@@ -0,0 +1,237 @@
+#include "../PluginStructs/P126_data_struct.h"
+
+#ifdef USES_P126
+
+// **************************************************************************/
+// Constructor
+// **************************************************************************/
+P126_data_struct::P126_data_struct(int8_t  dataPin,
+                                   int8_t  clockPin,
+                                   int8_t  latchPin,
+                                   uint8_t chipCount)
+  : _dataPin(dataPin), _clockPin(clockPin), _latchPin(latchPin), _chipCount(chipCount) {
+  shift = new (std::nothrow) ShiftRegister74HC595_NonTemplate(_chipCount, _dataPin, _clockPin, _latchPin);
+}
+
+// **************************************************************************/
+// Destructor
+// **************************************************************************/
+P126_data_struct::~P126_data_struct() {
+  if (nullptr != shift) {
+    delete shift;
+    shift = nullptr;
+  }
+}
+
+const uint32_t P126_data_struct::getChannelState(uint8_t offset, uint8_t size) const {
+  uint32_t result       = 0u;
+  const uint8_t *pvalue = shift->getAll();
+  const uint8_t  last   = offset + size;
+
+  if (nullptr != pvalue) {
+    uint16_t sft = 0u;
+
+    for (uint8_t ofs = offset; ofs < last; ofs++, sft++) {
+      result += (pvalue[ofs] << (8 * sft));
+    }
+
+    # ifdef P126_DEBUG_LOG
+
+    if (loglevelActiveFor(LOG_LEVEL_DEBUG)) {
+      String log = F("74hc595: getChannelState offset: ");
+      log += offset;
+      log += F(", size: ");
+      log += size;
+      log += F(", result: ");
+      log += result;
+      log += F("/0x");
+      log += String(result, HEX);
+      addLog(LOG_LEVEL_DEBUG, log);
+    }
+    # endif // ifdef P126_DEBUG_LOG
+  }
+  return result;
+}
+
+bool P126_data_struct::plugin_read(struct EventStruct *event) {
+  const uint16_t last = P126_CONFIG_SHOW_OFFSET + (VARS_PER_TASK * 4);
+  uint8_t varNr       = 0;
+
+  for (uint16_t index = P126_CONFIG_SHOW_OFFSET; index < _chipCount && index < last && varNr < VARS_PER_TASK; index += 4, varNr++) {
+    uint32_t result = getChannelState(index, min(VARS_PER_TASK, _chipCount - index));
+    UserVar.setUint32(event->TaskIndex, varNr, result);
+  }
+  return true;
+}
+
+bool P126_data_struct::plugin_write(struct EventStruct *event,
+                                    const String      & string) {
+  bool   success = false;
+  String command = parseString(string, 1);
+
+  if (command.startsWith(F("74hc"))) {
+    if (command.equals(F("74hcset")) || command.equals(F("74hcsetnoupdate"))) {
```

Comment:
@tonhuisman how about such idea : 
`bool hc_update = !command.contains(F("noupdate"));`
than:
`if (command.equals(F("74hcset")) || command.equals(F("74hcsetnoupdate")))` 
we can change to  
`if (command.startsWithF("74hcset"))`

and `shift->set(pin - 1, value, command.equals(F("74hcset")));` goes to `shift->set(pin - 1, value, hc_update);`

Reply:
> `if (command.equals(F("74hcset")) || command.equals(F("74hcsetnoupdate")))`
> we can change to
> `if (command.startsWithF("74hcset"))`

That's not really feasible as there are also the `74hcsetall` commands. Having a single bool for `noupdate` might save a few bytes, though. I'll improve that.

Reply:
@tonhuisman idea is to save more cpu time (string compare), you can use it also for 74hcsetall